### ISSUE TITLE
composer install w/o lock now triggers post-update-cmd

### ISSUE THREAD

**Issue by andrerom** [composer/composer#5128]
Using [beta2](https://github.com/composer/composer/releases/tag/1.0.0-beta2):

#### With the following composer.json:

```json
{
    "scripts": {
        "post-install-cmd": [
            "(..)ScriptHandler::dumpAssets"
        ],
        "post-update-cmd": [
        ]
    },
}
```

And no lock *(dev-master of meta project, where lock is counter productive, unlike with tags)*.

#### When I run this command:

```
composer install --no-progress --no-interaction --prefer-dist
# or
composer create-project (..)
```

#### This happens:
Nothing, no assets are dumped. Previously, install and create-project always triggered *install* scripts

#### And I expected this to happen:
Asset being dumped when I perform an install on a empty setup. However not on update as it is a somewhat slow process, and patch updates *(unlike major/minor **upgrades**)* usually don't have new assets.

Don't get me wrong, I perfectly understand why the semantic was changed, as install with and without lock is different in some ways, however install and update is from user and project perspective different as well.

TL;DR; `install with lock !== install without lock !== update`

**Comment by Seldaek** [comment]
I don't see how this makes sense to dump the assets only on install and not on update. If I run an update and get a new release with new assets, why would I not get the new assets dumped?

Previously you were working under broken assumptions IMO, and now you (and a few other projects) are forced to clean up and remove those assumptions from the workflow. I think that's ultimately a good thing, even if it includes a bit of pain along the way.

**Comment by andrerom** [comment]
Somewhat agree, it was like this for other reasons, no longer valid actually when checking them now. But while example is maybe not the best, I still feel this is something that can bite us down the line,  I'll try to think up a better example :)

**Comment by Seldaek** [comment]
Well until now everyone that came complaining has failed to convince me that this was a bad change ;) If you got more let me know though but as it stands I am fairly confident this won't be reverted because it makes things more uniform and predictable.

**Comment by andrerom** [comment]
Never asked for a revert, I think the change is good direction, just suggests there should be possible to separate the three cases in scripts.

Would for instance be enough if script event object has property containing original command executed by user. That would mean result of above with script added to `post-update-cmd` would be `$event->getName() === 'post-update-cmd' && $event->getCommand() === 'install' `

**Comment by Seldaek** [comment]
Hmm yes I see, not sure if there is a huge benefit though. Most things I can think of really belong in either clearly post-update, post-install or both, but I don't see the use case for "I run install but actually I have no lock file so eh run update but run my install scripts still". It's a kind of "please do what I think not what I say" which is not realistic with today's computing tech ;)

**Comment by andrerom** [comment]
Ok, what about a "post-composer-lock-creation" and "post-composer-lock-update" event then? :)

**Comment by stof** [comment]
@andrerom the rule is simple: only update can alter the lock file.
And why should something run only when a lock file is created for the first time ?

**Comment by andrerom** [comment]
> And why should something run only when a lock file is created for the first time ?
To be able to distinguish an install *(without lock)* from an update, script wise, as they are two separate things conceptually.

**Comment by Seldaek** [comment]
OK won't fix sounds good I guess :) We can always talk again if there's an actual use case.
